### PR TITLE
Add equality checks to `prefer-t-regex` rule

### LINKED ISSUE #275: Detect more assertions in the `t-prefer-regex` rule
[Issuehunt badges](https://issuehunt.io/r/avajs/eslint-plugin-ava/issues/275)

See https://github.com/avajs/eslint-plugin-ava/issues/158#issuecomment-523830714

---

> *Note: This issue has a bounty, so it's expected that you are an experienced programmer and that you give it your best effort if you intend to tackle this. Don't forget, if applicable, to add tests, docs (double-check for typos). And don't be sloppy. Review your own diff multiple times and try to find ways to improve and simplify your code. Instead of asking too many questions, present solutions. The point of an issue bounty is to reduce my workload, not give me more. Include a 🦄 in your PR description to indicate that you've read this. Thanks for helping out 🙌* - @sindresorhus

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $40.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($40.00)

### Submitted pull Requests
- [#297 Add equality checks to `prefer-t-regex` rule](https://issuehunt.io/r/avajs/eslint-plugin-ava/pull/297)
- [#319 Detect more assertions in `prefer-t-regex`](https://issuehunt.io/r/avajs/eslint-plugin-ava/pull/319)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/avajs/eslint-plugin-ava/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

### PR DESCRIPTION
**Pull request by JaniM**
Fixes #275.

These are the maximal reasonable rules I could think of.

image

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#275: Detect more assertions in the `t-prefer-regex` rule](https://issuehunt.io/repos/49722492/issues/275)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

File: test/prefer-t-regex.js [hunk 1]
```diff
@@ -62,6 +62,26 @@ ruleTester.run('prefer-t-regex', rule, {
 			code: header + 'const reg = /\\d+/;\ntest(t => t.true(reg.test(foo.bar())));',
 			output: header + 'const reg = /\\d+/;\ntest(t => t.regex(foo.bar(), reg));',
 			errors: errors('regex')
+		},
+		{
+			code: header + 'test(t => t.is(foo(), /\\d+/));',
```
Comment (sindresorhus) [inline test/prefer-t-regex.js]: Can you also add version of these tests using `new RegExp()`?

File: rules/prefer-t-regex.js [hunk 2]
```diff
@@ -13,75 +13,134 @@ const create = context => {
...
+			message: `Prefer using the \`t.${assertion}()\` assertion.`,
+			fix
+		});
+	};
+
+	const equalityHandler = node => {
+		const firstArg = node.arguments[0];
+		const secondArg = node.arguments[1];
+
+		const firstIsRx = isRegExp(firstArg);
```
Comment (sindresorhus) [inline rules/prefer-t-regex.js]: Don't use abbreviations.

```suggestion
		const firstArgumentIsRegex = isRegExp(firstArgument);
```

File: rules/prefer-t-regex.js [hunk 3]
```diff
@@ -13,75 +13,134 @@ const create = context => {
...
+		context.report({
+			node,
+			message: `Prefer using the \`t.${assertion}()\` assertion.`,
+			fix
+		});
+	};
+
+	const equalityHandler = node => {
+		const firstArg = node.arguments[0];
+		const secondArg = node.arguments[1];
```
Comment (sindresorhus) [inline rules/prefer-t-regex.js]:
```suggestion
		const [firstArg, secondArg] = node.arguments;
```

File: test/prefer-t-regex.js (hunk 1 above)
Comment (JaniM) [reply test/prefer-t-regex.js]: Do we want to expand the functionality for this? The original regex detection logic only matches regex literals, not expressions resolving to a regex. Resolving whether or not any given expression resolves to a regex pretty much requires eval, so we'd have to special-case `new RegExp()`.

File: rules/prefer-t-regex.js (hunk 2 above)
Comment (JaniM) [reply rules/prefer-t-regex.js]: Fixed.

File: rules/prefer-t-regex.js (hunk 3 above)
Comment (JaniM) [reply rules/prefer-t-regex.js]: Fixed.

File: test/prefer-t-regex.js (hunk 1 above)
Comment (sindresorhus) [reply test/prefer-t-regex.js]: We don't need to resolve expression, but we should support an inline `new RegExp()` call.

**Comment by JaniM** [comment]
Due to the current situation, I don't have time to finish this. Anyone can pick this up.

**Comment by novemberborn** [comment]
Thanks for the update @JaniM. Good luck ❤️

**Comment by sindresorhus** [comment]
https://github.com/avajs/eslint-plugin-ava/issues/275#issuecomment-647293780

**Pull request closed without merging**